ls8/cpu.py

```diff
@@ -1,42 +1,113 @@
 """CPU functionality."""
 
-import sys
+
+ldi = 0b10000010
+prn = 0b01000111
+hlt = 0b00000001
+mul = 0b10100010
+push = 0b01000101
+pop = 0b01000110
+
 
 class CPU:
     """Main CPU class."""
 
     def __init__(self):
         """Construct a new CPU."""
-        pass
+        self.ram = [0] * 256
 
-    def load(self):
-        """Load a program into memory."""
+        self.reg = [0] * 8
+        self.pc = 0
+
+        self.sp = 7
+        self.running = False
+
+        self.reg[self.sp] = 0xF4
+
+        self.branchtable = {}
+        self.branchtable[ldi] = self.ldi
+        self.branchtable[prn] = self.prn
+        self.branchtable[hlt] = self.hlt
+        self.branchtable[mul] = self.mul
+        self.branchtable[push] = self.push
+        self.branchtable[pop] = self.pop
+
+    def ram_read(self, mar):
+        return self.ram[mar]
+
+    def ram_write(self, mar, mdr):
+        self.ram[mar] = mdr
+
+    def hlt(self):
+        self.running = False
+
+    def ldi(self):
+        print("ldi")
+        operand_a = self.ram_read(self.pc + 1)
+        operand_b = self.ram_read(self.pc + 2)
+
+        self.reg[operand_a] = operand_b
+        self.pc += 3
+
+    def prn(self):
+        print("prn")
+        operand_a = self.ram_read(self.pc + 1)
 
-        address = 0
+        print(self.reg[operand_a])
+        self.pc += 2
 
-        # For now, we've just hardcoded a program:
+    def mul(self):
+        print("mul")
+        operand_a = self.ram_read(self.pc + 1)
+        operand_b = self.ram_read(self.pc + 2)
 
-        program = [
-            # From print8.ls8
-            0b10000010, # LDI R0,8
-            0b00000000,
-            0b00001000,
-            0b01000111, # PRN R0
-            0b00000000,
-            0b00000001, # HLT
-        ]
+        self.reg[operand_a] * self.reg[operand_b]
+        self.pc += 3
 
-        for instruction in program:
-            self.ram[address] = instruction
-            address += 1
+    def push(self):
+        print("push")
 
+        reg = self.ram[self.pc + 1]
+        val = self.reg[reg]
+        self.reg[self.sp] -= 1
+
+        self.ram[self.reg[self.sp]] = val
+        self.pc += 2
+
+    def pop(self):
+        print("pop")
+        reg = self.ram[self.pc + 1]
+        val = self.ram[self.reg[self.sp]]
+        self.reg[reg] = val
+        self.reg[self.sp] += 1
+        self.pc += 2
```

Comment:
Nice! Getting the value from memory. Storing the value from the stack in the register 👍